File: 876-middle-of-the-linked-list/876-middle-of-the-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution 
{
public:
    ListNode* middleNode(ListNode* head) 
    {
        // Traverses the list with two pointers to determine the middle node of the linked list
        ListNode* curr = head;
        ListNode* mid = head;
        int midIdx = 0;
        int currSize = 0;
        while(curr != nullptr)
        {
            currSize++;
            curr = curr->next;
            
            if(midIdx < currSize / 2)
            {
                midIdx++;
                mid = mid->next;
            }
        }
        
        return mid;
    }
};

File: 876-middle-of-the-linked-list/876-middle-of-the-linked-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution 
{
    public ListNode MiddleNode(ListNode head) 
    {
        ListNode curr = head;
        int len = 0;
        int mid = 0;
        while(head != null)
        {
            if(len / 2 > mid)
            {
                curr = curr.next;
                mid++;
            }
            
            head = head.next;
            len++;
        }
        
        return len % 2 == 0 ? curr.next : curr;
    }
}